### PR TITLE
tools: first version of mei40To50.xsl

### PR DESCRIPTION
**Pull request by kepper**
seems to properly work on files from the sample collection

### CODE REVIEW

**Comment by bwbohl** [comment]
please also implement

~~- [ ] The values "italic" and "oblique" duplicate values available in data.FONTSTYLE. Fixes #1102https://github.com/music-encoding/music-encoding/pull/1188~~

**Comment by bwbohl** [comment]
@musicEnfanthen is this also tested in the docker-mei container?

**Comment by kepper** [comment]
> please also implement
> 
>     * [ ]  The values "italic" and "oblique" duplicate values available in data.FONTSTYLE. Fixes #1102[Remove "italic" and "oblique" values from data.TEXTRENDITIONLIST music-encoding#1188](https://github.com/music-encoding/music-encoding/pull/1188)

That seems to open a whole other can of worms, as `@rend` still contains other values like "bold" (which could go into `@fontweight`), so this whole change seems half-baked. Are we really clear that this is the way to go for a v5 release?

**Comment by bwbohl** [comment]
> > please also implement
> > ```
> > * [ ]  The values "italic" and "oblique" duplicate values available in data.FONTSTYLE. Fixes #1102[Remove "italic" and "oblique" values from data.TEXTRENDITIONLIST music-encoding#1188](https://github.com/music-encoding/music-encoding/pull/1188)
> > ```
> 
> That seems to open a whole other can of worms, as `@rend` still contains other values like "bold" (which could go into `@fontweight`), so this whole change seems half-baked. Are we really clear that this is the way to go for a v5 release?

not applicable anymore as we undid the merge, cf. https://github.com/music-encoding/music-encoding/pull/1188#issuecomment-1592045421

**Comment by musicEnfanthen** [comment]
Open issues:
- [x] key.sig --> keysig (see music-encoding/music-encoding#933)
- [x] "instrDef/@n" in from att.nNumberlike to att.nInterger (see music-encoding/music-encoding#933)
- [x] keysig.show --> keysig.visible
- [x] keysig.showchange --> keysig.cancelaccid (as the datatype changed from boolean to none, before, after, before-bar the values should be transfromed as follows: false --> none; true --> before [with a message that indicates the possibility of other values]
- [x] the XSLT assumes there is already an `<appInfo>` element to which it can add it's own information; if `<appInfo>` is not there yet, it should be inserted. same is true for `revisionDesc`

for more please also refer to: #31

File: mei40To50/mei40To50.xsl [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+    <xsl:variable name="nl">
+        <xsl:text>&#xa;</xsl:text>
+    </xsl:variable>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>MEI version</xd:p>
+        </xd:desc>
+    </xd:doc>
+    <xsl:variable name="meiversion" select="'5.0.0'"/>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
    <xsl:variable name="meiversion" select="'5.0'"/>
```

File: mei40To50/mei40To50.xsl [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+            <!-- Add a record of the conversion to revisionDesc -->
+            <change xmlns="http://www.music-encoding.org/ns/mei">
+                <xsl:if test="count(mei:change[@n]) = count(mei:change)">
+                    <xsl:attribute name="n" select="count(mei:change) + 1"/>
+                </xsl:if>
+                <xsl:attribute name="resp">
+                    <xsl:value-of select="concat('#', $progid)"/>
+                </xsl:attribute>
+                <changeDesc>
+                    <p><xsl:value-of select="'Converted to MEI version 5.0.0 using ' || $progname || ', version ' || $version"/></p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
                    <p><xsl:value-of select="'Converted to MEI version 5.0 using ' || $progname || ', version ' || $version"/></p>
```

**Comment by bwbohl** [comment]
For the record: most of the issues mentioned in [above](https://github.com/music-encoding/encoding-tools/pull/30#issuecomment-1633148456) are already fixed on my local clone

**Comment by bwbohl** [comment]
before proceeding though, I would like some feedback on https://github.com/music-encoding/sample-encodings/issues/77

File: mei40To50/mei40To50.xsl [hunk 3]
```diff
@@ -0,0 +1,718 @@
...
+            <xd:p>Removes @visible on mRest</xd:p>
+        </xd:desc>
+    </xd:doc>
+    <xsl:template match="mei:mRest/@visible">
+        <xsl:if test="$verbose">
+            <xsl:message select="'Dropping @visible on mRest ' || ancestor-or-self::mei:*[@xml:id][1]/@xml:id"/>
+        </xsl:if>
+    </xsl:template>
+    
+   <xd:doc>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
    <xd:doc>
```

File: mei40To50/mei40To50.xsl [hunk 4]
```diff
@@ -0,0 +1,718 @@
...
+                <xsl:value-of select="'Changing @sig.showchange to @cancelaccid on ' || local-name(parent::mei:*) || ' ' || ancestor-or-self::mei:*[@xml:id][1]/@xml:id || '. '"/>
+                <xsl:text>Please note: converting value from </xsl:text>
+                <xsl:value-of select="$value.old"/><xsl:text> to </xsl:text><xsl:value-of select="$value.new"/><xsl:text>. </xsl:text>
+                <xsl:if test="$value.old = 'true'">There are several alternatives for encoding the style of MEI 4 @sig.showchange='true' in MEI 5.0 (before, after, before-bar), defaulting to 'before'.</xsl:if>
+            </xsl:message>
+        </xsl:if>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>Resolve changes in @meter.form, moving @meter.form="invis" to @meter.visible</xd:desc>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
        <xd:desc>Resolve changes in @meter.form, moving @meter.form="invis" to @meter.visible="false".</xd:desc>
```

File: mei40To50/mei40To50.xsl [hunk 5]
```diff
@@ -0,0 +1,718 @@
...
+    <xsl:template match="@text.dist">
+        <xsl:attribute name="dir.dist" select="."/>
+        <xsl:if test="$verbose">
+            <xsl:message select="'Replacing @text.dist with @dir.dist on ' || local-name(parent::mei:*) || ' ' || ancestor-or-self::mei:*[@xml:id][1]/@xml:id || '. Alternatives could be @reh.dist or @tempo.dist.'"/>
+        </xsl:if>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>Adjust value of MIDI name</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Replace a value of "Bagpipe" on @midi.instrname with "Bag_pipe".</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+            <xsl:variable name="types" select="(tokenize(normalize-space(@type), ' '), 'fingerprint')" as="xs:string*"/>
+            <xsl:attribute name="type" select="string-join($types, ' ')"/>
+            <xsl:apply-templates select="@* except @type"/>
+            <xsl:apply-templates select="node()"/>
+        </identifier>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>Decide whether to keep instrDef/@n or replace it with instrDef/@label, depending on its contents.</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Move a non-numeric value on instrDef/@n to instrDef/@label.</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 7]
```diff
@@ -0,0 +1,718 @@
...
+        </xsl:copy>
+        
+        <xsl:if test="$verbose">
+            <xsl:message select="'Separating letter-spacing and line-height into separate attributes on ' || local-name(parent::mei:*) || ' ' || ancestor-or-self::mei:*[@xml:id][1]/@xml:id"/>
+        </xsl:if>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>Removes fingerprint in favor of identifier, which is allowed everywhere where fingerprint was allowed.</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Replace deprecated fingerprint in favor of identifier/@type="fingerprint", which is allowed everywhere where fingerprint was allowed.</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 8]
```diff
@@ -0,0 +1,718 @@
...
+                <xsl:message terminate="yes">The @n on instrDef seems to be invalid in the source file.
+                <xsl:value-of select="document-uri(root())"/>
+                </xsl:message>
+            </xsl:otherwise>
+        </xsl:choose>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>A value of "dblwhole" on @head.mod replaced with "fences".</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Replace a value of "dblwhole" on @head.mod with "fences".</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 9]
```diff
@@ -0,0 +1,718 @@
...
+            </xsl:when>
+            <xsl:otherwise>
+                <xsl:copy-of select="."/>
+            </xsl:otherwise>
+        </xsl:choose>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>Resolve @text.dist</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Resolve @text.dist to @dir.dist.</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 10]
```diff
@@ -0,0 +1,718 @@
...
+    </xd:doc>
+    <xsl:template match="mei:mSpace/@instr">
+        <xsl:if test="$verbose">
+            <xsl:message select="'Dropping @instr on mSpace ' || ancestor-or-self::mei:*[@xml:id][1]/@xml:id"/>
+        </xsl:if>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>Removes @instr on multiRest</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Remove @instr on multiRest.</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 11]
```diff
@@ -0,0 +1,718 @@
...
+    </xd:doc>
+    <xsl:template match="mei:mRest/@instr">
+        <xsl:if test="$verbose">
+            <xsl:message select="'Dropping @instr on mRest ' || ancestor-or-self::mei:*[@xml:id][1]/@xml:id"/>
+        </xsl:if>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>Removes @instr on mSpace</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Remove @instr on mSpace.</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 12]
```diff
@@ -0,0 +1,718 @@
...
+    </xd:doc>
+    <xsl:template match="mei:mRest/@visible">
+        <xsl:if test="$verbose">
+            <xsl:message select="'Dropping @visible on mRest ' || ancestor-or-self::mei:*[@xml:id][1]/@xml:id"/>
+        </xsl:if>
+    </xsl:template>
+    
+   <xd:doc>
+        <xd:desc>
+            <xd:p>Removes @instr on mRest</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Remove @instr on mRest.</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 13]
```diff
@@ -0,0 +1,718 @@
...
+        </xsl:if>
+        <xsl:copy>
+            <xsl:attribute name="func" select="'first'"/>
+            <xsl:apply-templates select="node() | @*"/>
+        </xsl:copy>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>Removes @visible on mRest</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Remove @visible on mRest.</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 14]
```diff
@@ -0,0 +1,718 @@
...
+                    <xsl:message select="'Inserting @meiversion on ' || local-name(parent::mei:*) || ' with value: ' || $meiversion"/>
+                </xsl:if>
+            </xsl:if>
+            <xsl:apply-templates select="node() | @*"/>
+        </xsl:copy>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>pgHead2 for subsequent pages is now encoded as pgHead with func="all", assuming another pgHead will use func="first"</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Replace pgHead2 for subsequent pages with pgHead/@func="all", assuming another pgHead will use @func="first".</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 15]
```diff
@@ -0,0 +1,718 @@
...
+            <xsl:message select="'pgHead2 ' || ancestor-or-self::mei:*[@xml:id][1]/@xml:id || ' changed to pgHead.'"/>
+        </xsl:if>
+        <pgHead xmlns="http://www.music-encoding.org/ns/mei" func="all">
+            <xsl:apply-templates select="node() | @*"/>
+        </pgHead>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>pgHead is now encoded as pgHead with func="first", assuming another pgHead2 was there and is now encoded as pgHead with func="all"</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Replace pgHead with pgHead/@func="first", assuming another pgHead2 was there and is now encoded as pgHead with @func="all".</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 16]
```diff
@@ -0,0 +1,718 @@
...
+        </xsl:if>
+        <xsl:copy>
+            <xsl:attribute name="func" select="'first'"/>
+            <xsl:apply-templates select="node() | @*"/>
+        </xsl:copy>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>pgFoot2 for subsequent pages is now encoded as pgFoot with func="all", assuming another pgFoot will use func="first"</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Replace pgFoot2 for subsequent pages with pgFoot/@func="all", assuming another pgFoot will use @func="first".</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 17]
```diff
@@ -0,0 +1,718 @@
...
+            <xsl:message select="'pgFoot2 ' || ancestor-or-self::mei:*[@xml:id][1]/@xml:id || ' changed to pgFoot.'"/>
+        </xsl:if>
+        <pgFoot xmlns="http://www.music-encoding.org/ns/mei" func="all">
+            <xsl:apply-templates select="node() | @*"/>
+        </pgFoot>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>pgFoot is now encoded as pgFoot with func="first", assuming another pgFoot2 was there and is now encoded as pgFoot with func="all"</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Replace pgFoot with pgFoot/@func="first", assuming another pgFoot2 was there and is now encoded as pgFoot with @func="all".</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 18]
```diff
@@ -0,0 +1,718 @@
...
+    </xd:doc>
+    <xsl:template match="mei:multiRest/@instr">
+        <xsl:if test="$verbose">
+            <xsl:message select="'Dropping @instr on multiRest ' || ancestor-or-self::mei:*[@xml:id][1]/@xml:id"/>
+        </xsl:if>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>Removes @instr on rest</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Remove @instr on rest.</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 19]
```diff
@@ -0,0 +1,718 @@
...
+                </xsl:if>
+            </xsl:when>
+            <xsl:otherwise>
+                <xsl:copy-of select="."/>
+            </xsl:otherwise>
+        </xsl:choose>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>Resolve changes in meterSig/@form, moving @form="invis" to @visible="false"</xd:desc>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
        <xd:desc>Resolve changes in meterSig/@form, moving @form="invis" to @visible="false".</xd:desc>
```

File: mei40To50/mei40To50.xsl [hunk 20]
```diff
@@ -0,0 +1,718 @@
...
+    <xsl:template match="@midi.instrname">
+        <xsl:attribute name="midi.instrname" select="replace(., 'Bagpipe', 'Bag_pipe')"/>
+        <xsl:if test="$verbose">
+            <xsl:message select="'Changing Bagpipe to Bag_pipe on @midi.instrname on ' || local-name(parent::mei:*) || ' ' || ancestor-or-self::mei:*[@xml:id][1]/@xml:id"/>
+        </xsl:if>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>Move letter-spacing and line-height on @rend to separate attributes</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Move letter-spacing and line-height on @rend to separate attributes.</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 21]
```diff
@@ -0,0 +1,718 @@
...
+    </xsl:template>
+    
+    
+    <!-- ======================================================================= -->
+    <!-- COPY TEMPLATE                                                           -->
+    <!-- ======================================================================= -->
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>Copy template</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Copy template.</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 22]
```diff
@@ -0,0 +1,718 @@
...
+    </xd:doc>
+    <xsl:template match="mei:revisionDesc">
+        <xsl:copy>
+            <xsl:apply-templates select="node() | @*"/>
+            <xsl:call-template name="revisionDesc-insert-change"/>
+        </xsl:copy>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>Add a record of the conversion to revisionDesc</xd:desc>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
        <xd:desc>Add a record of the conversion to revisionDesc.</xd:desc>
```

File: mei40To50/mei40To50.xsl [hunk 23]
```diff
@@ -0,0 +1,718 @@
...
+        </xsl:choose>
+    </xsl:template>
+    
+    <!-- ======================================================================= -->
+    <!-- SELF-DOCUMENTATION TEMPLATES                                            -->
+    <!-- ======================================================================= -->
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>Add a change element for the conversion</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Add a change element for the conversion.</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 24]
```diff
@@ -0,0 +1,718 @@
...
+    </xd:doc>
+    <xsl:variable name="meiversion" select="'5.0'"/>
+    
+    <!-- ======================================================================= -->
+    <!-- MAIN OUTPUT TEMPLATE                                                    -->
+    <!-- ======================================================================= -->
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p>Start template</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl]:
```suggestion
            <xd:p>Start template.</xd:p>
```

File: mei40To50/mei40To50.xsl [hunk 25]
```diff
@@ -0,0 +1,718 @@
...
+    xmlns:xs="http://www.w3.org/2001/XMLSchema"
+    xmlns:math="http://www.w3.org/2005/xpath-functions/math"
+    xmlns:xd="http://www.oxygenxml.com/ns/doc/xsl"
+    xmlns:mei="http://www.music-encoding.org/ns/mei"
+    exclude-result-prefixes="xs math xd mei"
+    version="3.0">
+    <xd:doc scope="stylesheet">
+        <xd:desc>
+            <xd:p><xd:b>Created on:</xd:b> May 25, 2023</xd:p>
+            <xd:p><xd:b>Author:</xd:b> Johannes Kepper</xd:p>
```
Comment (musicEnfanthen) [inline mei40To50/mei40To50.xsl:12]:
```suggestion
            <xd:p><xd:b>Author:</xd:b> Johannes Kepper</xd:p>
            <xd:p><xd:b>Author:</xd:b> Benjamin W. Bohl</xd:p>
```

**Review by musicEnfanthen — COMMENTED** [review]
Some suggestions to unify spelling and syntax of docs

**Review by bwbohl — APPROVED** [review]

**Pull request merged**